src/Workspaces/CoreTest/SolutionTests/SolutionWithSourceGeneratorTests.cs

```diff
@@ -0,0 +1,139 @@
+
+            var documentId = project.DocumentIds.Single();
+
+            await AssertCompilationContainsOneRegularAndOneGeneratedFile(project, documentId);
+
+            project = project.Solution.WithDocumentText(documentId, SourceText.From("// Changed Source File")).Projects.Single();
+
+            await AssertCompilationContainsOneRegularAndOneGeneratedFile(project, documentId);
+
+            static async Task AssertCompilationContainsOneRegularAndOneGeneratedFile(Project project, DocumentId documentId)
+            {
+                var compilation = await project.GetRequiredCompilationAsync(CancellationToken.None);
+
+                var regularDocumentSyntaxTree = await project.GetRequiredDocument(documentId).GetRequiredSyntaxTreeAsync(CancellationToken.None);
+                Assert.Contains(regularDocumentSyntaxTree, compilation.SyntaxTrees);
+
+                var generatedSyntaxTree = Assert.Single(compilation.SyntaxTrees.Where(t => t != regularDocumentSyntaxTree));
+                Assert.Null(project.GetDocument(generatedSyntaxTree));
+            }
+        }
+
+        private sealed class TestGeneratorReference : AnalyzerReference
+        {
+            private readonly ISourceGenerator _generator;
+
+            public TestGeneratorReference(ISourceGenerator generator)
+            {
+                _generator = generator;
+            }
+
+            public override string? FullPath => null;
+            public override object Id => this;
+
+            public override ImmutableArray<DiagnosticAnalyzer> GetAnalyzers(string language) => ImmutableArray<DiagnosticAnalyzer>.Empty;
+            public override ImmutableArray<DiagnosticAnalyzer> GetAnalyzersForAllLanguages() => ImmutableArray<DiagnosticAnalyzer>.Empty;
+            public override ImmutableArray<ISourceGenerator> GetGenerators() => ImmutableArray.Create(_generator);
+        }
+
+        // PROTOTYPE: find a way to reuse this implementation from the compiler unit tests
+        private sealed class AdditionalFileAddedGenerator : ISourceGenerator
+        {
+            public bool CanApplyChanges { get; set; } = true;
+
+            public void Execute(SourceGeneratorContext context)
+            {
+                foreach (var file in context.AdditionalFiles)
+                {
+                    AddSourceForAdditionalFile(context, file);
+                }
+            }
+
+            public void Initialize(InitializationContext context)
+            {
+                // PROTOTYPE: context.RegisterForAdditionalFileChanges(UpdateContext);
+            }
+
+            private void AddSourceForAdditionalFile(SourceGeneratorContext context, AdditionalText file) => context.AddSource(GetGeneratedFileName(file.Path), SourceText.From("", Encoding.UTF8));
+
+            private string GetGeneratedFileName(string path) => $"{Path.GetFileNameWithoutExtension(path)}.generated";
+        }
+    }
+}
```

Comment:
i appreciate test.  one thing that we may need to invest in IMO is a way to do fine-grained compilation tracker testing. i.e to actually validate that we are or aren't doing specific pieces of work at specific times.